keysync_initiator.go

```diff
@@ -0,0 +1,172 @@
+	// supposed to use.
+	ourAttDoc, err := attest(theirNonce[:], ourNonce[:], sbKey.Bytes())
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Send our attestation document to the remote enclave, and get theirs in
+	// return.
+	theirAttDoc, err := requestAttDoc(addr, ourAttDoc)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Finally, verify the attestation document and extract the key material.
+	if err := processAttDoc(theirAttDoc, &ourNonce, sbKey, keyMaterial); err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	return nil
+}
+
+// requestNonce requests a nonce from the remote enclave specified by 'addr'.
+func requestNonce(addr string) (nonce, error) {
+	errStr := "failed to fetch nonce from remote enclave"
+
+	endpoint := fmt.Sprintf("%s%s", addr, pathNonce)
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Decode the Base64-encoded nonce.
+	raw, err := base64.StdEncoding.DecodeString(strings.TrimSpace(string(body)))
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	if len(raw) != nonceLen {
+		return nonce{}, errors.New("remote enclave's nonce has incorrect length")
+	}
+	var n nonce
+	copy(n[:], raw)
+	return n, nil
+}
+
+// requestAttDoc takes as input the remote enclave's address (e.g.,
+// <https://example.com>) and our attestation document.  The function then
+// submits our attestation document to the remote enclave, and returns the
+// remote enclave's attestation document.
+func requestAttDoc(addr string, ourAttDoc []byte) ([]byte, error) {
+	errStr := "failed to fetch attestation doc from remote enclave"
+
+	endpoint := fmt.Sprintf("%s%s", addr, pathKeys)
+
+	// Finally, send our attestation document to the remote enclave.  If
+	// everything works out, the remote enclave is going to respond with its
+	// attestation document.
+	b64AttDoc := base64.StdEncoding.EncodeToString(ourAttDoc)
+	resp, err := http.Post(
+		endpoint,
+		"application/octet-stream",
+		bytes.NewBufferString(b64AttDoc),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", errStr, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	theirAttDoc, err := base64.StdEncoding.DecodeString(string(body))
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	return theirAttDoc, nil
+}
+
+// processAttDoc first verifies that the remote enclave's attestation document
+// is authentic, and then attempts to decrypt and extract the key material that
+// the remote enclave provided in its attestation document.
+func processAttDoc(
+	theirAttDoc []byte,
+	ourNonce *nonce,
+	sbKey *sbKey,
+	keyMaterial any,
+) error {
```

Comment:
Would it make sense to return `(keyMaterial, error)` like the other functions, instead of using an out parameter? Are there problems with it being an `any`? Maybe that's what 'write to an interface' means...

Reply:
We could, but that would be redundant: We need the `keyMaterial` argument to tell the JSON decoder where to write the decoded JSON. After the decoding, we could return `(keyMaterial, error)` as you suggested but that strikes me as unnecessary because we might as well write the decoded JSON directly to the provided argument.

Reply:
Ok. In Rust (and sometimes C++) the compiler optimizes return by value into the same thing as out-parameter use. So you can write to and return a temporary with no copy overhead to make the function's use more functional.

If that's not normal go style, I have no objection to reflecting the json api's convention.